{
    "ID": {
        "GitHub_Advisory_Database": {
            "aliases": [
                "CVE-2022-21728"
            ]
        },
        "Nomisec": {
            "id": 486782625,
            "name": "CVE-2022-21728-test",
            "full_name": "mwina/CVE-2022-21728-test"
        }
    },
    "summary": "Out of bounds read in Tensorflow",
    "details": {
        "GitHub_Advisory_Database": "### Impact \nThe [implementation of shape inference for `ReverseSequence`](https://github.com/tensorflow/tensorflow/blob/5100e359aef5c8021f2e71c7b986420b85ce7b3d/tensorflow/core/ops/array_ops.cc#L1636-L1671) does not fully validate the value of `batch_dim` and can result in a heap OOB read:\n\n```python\nimport tensorflow as tf\n\n@tf.function\ndef test():\n  y = tf.raw_ops.ReverseSequence(\n    input = ['aaa','bbb'],\n    seq_lengths = [1,1,1],\n    seq_dim = -10,\n    batch_dim = -10 )\n  return y\n    \ntest()\n```\n\nThere is a check to make sure the value of `batch_dim` does not go over the rank of the input, but there is no check for negative values:\n\n```cc\n  const int32_t input_rank = c->Rank(input);\n  if (batch_dim >= input_rank) {\n    return errors::InvalidArgument( \n        \"batch_dim must be < input rank: \", batch_dim, \" vs. \", input_rank);\n  }\n  // ...\n  \n  DimensionHandle batch_dim_dim = c->Dim(input, batch_dim);\n``` \n    \nNegative dimensions are allowed in some cases to mimic Python's negative indexing (i.e., indexing from the end of the array), however if the value is too negative then [the implementation of `Dim`](https://github.com/tensorflow/tensorflow/blob/5100e359aef5c8021f2e71c7b986420b85ce7b3d/tensorflow/core/framework/shape_inference.h#L415-L428) would access elements before the start of an array:\n\n```cc\n  DimensionHandle Dim(ShapeHandle s, int64_t idx) {\n    if (!s.Handle() || s->rank_ == kUnknownRank) {\n      return UnknownDim();\n    }\n    return DimKnownRank(s, idx);\n  } \nÂ·\n  static DimensionHandle DimKnownRank(ShapeHandle s, int64_t idx) {\n    CHECK_NE(s->rank_, kUnknownRank);\n    if (idx < 0) {\n      return s->dims_[s->dims_.size() + idx];\n    }\n    return s->dims_[idx];\n  }\n```\n\n### Patches\nWe have patched the issue in GitHub commit [37c01fb5e25c3d80213060460196406c43d31995](https://github.com/tensorflow/tensorflow/commit/37c01fb5e25c3d80213060460196406c43d31995).\n\nThe fix will be included in TensorFlow 2.8.0. We will also cherrypick this commit on TensorFlow 2.7.1, TensorFlow 2.6.3, and TensorFlow 2.5.3, as these are also affected and still in supported range.\n\n### For more information\nPlease consult [our security guide](https://github.com/tensorflow/tensorflow/blob/master/SECURITY.md) for more information regarding the security model and how to contact us with issues and questions.\n\n### Attribution\nThis vulnerability has been reported by Yu Tian of Qihoo 360 AIVul Team."
    },
    "description": {
        "Nomisec": null
    },
    "topics": [
        "tensorflow",
        "return",
        "github",
        "input",
        "cc",
        "blob",
        "commit",
        "com",
        "https",
        "value"
    ],
    "severity": [],
    "affected": [],
    "references": [
        {
            "type": "WEB",
            "url": "https://github.com/tensorflow/tensorflow/security/advisories/GHSA-6gmv-pjp9-p8w8"
        },
        {
            "type": "ADVISORY",
            "url": "https://nvd.nist.gov/vuln/detail/CVE-2022-21728"
        },
        {
            "type": "WEB",
            "url": "https://github.com/tensorflow/tensorflow/commit/37c01fb5e25c3d80213060460196406c43d31995"
        },
        {
            "type": "WEB",
            "url": "https://github.com/tensorflow/tensorflow"
        },
        {
            "type": "WEB",
            "url": "https://github.com/tensorflow/tensorflow/blob/5100e359aef5c8021f2e71c7b986420b85ce7b3d/tensorflow/core/framework/shape_inference.h#L415-L428"
        },
        {
            "type": "WEB",
            "url": "https://github.com/tensorflow/tensorflow/blob/5100e359aef5c8021f2e71c7b986420b85ce7b3d/tensorflow/core/ops/array_ops.cc#L1636-L1671"
        }
    ],
    "proof_of_concept": {
        "owner": {
            "html_url": "https://github.com/mwina"
        },
        "html_url": "https://github.com/mwina/CVE-2022-21728-test"
    },
    "has_discussions": null,
    "database_specific": {
        "cwe_ids": [
            "CWE-125"
        ],
        "severity": "HIGH",
        "github_reviewed": true,
        "github_reviewed_at": "2022-02-03T18:29:10Z",
        "nvd_published_at": "2022-02-03T11:15:00Z"
    },
    "Time_Description": {
        "GitHub_Advisory_Database": {
            "modified": "2022-02-09T18:29:24Z",
            "published": "2022-02-09T18:29:24Z",
            "withdrawn": null
        },
        "Nomisec": {
            "created_at": "2022-04-28T23:56:58Z",
            "updated_at": "2022-04-29T00:00:12Z",
            "pushed_at": "2022-04-29T00:36:23Z"
        }
    }
}